zetaclient/hsm_signer.go

```diff
@@ -0,0 +1,167 @@
+func GenerateKey(label string, algorithm keystone.KeygenAlgorithm, config *crypto11.Config) (*keystone.CryptoKey, error) {
+	keyring, err := keystone.NewPkcs11(config)
+	if err != nil {
+		return nil, err
+	}
+	return keyring.NewKey(algorithm, label)
+}
+
+// GetHSMAddress This address is generated by secp256k1 curve from cosmos sdk
+func GetHSMAddress(config *crypto11.Config, label string) (types.Address, types.PubKey, error) {
+	keyring, err := keystone.NewPkcs11(config)
+	if err != nil {
+		return nil, nil, err
+	}
+	key, err := keyring.Key(label)
+	if err != nil {
+		return nil, nil, err
+	}
+	pubKey := key.PubKey()
+	return pubKey.Address(), pubKey, nil
+}
+
+// SignWithHSM signs a given tx with a named key.
+// This is adapted from github.com/cosmos/cosmos-sdk/client/tx Sign() function; Modified to use an HSM.
+// The resulting signature will be added to the transaction builder overwriting the previous
+// ones if overwrite=true (otherwise, the signature will be appended).
+// Signing a transaction with multiple signers in the DIRECT mode is not supported and will
+// return an error.
+// An error is returned upon failure.
+func SignWithHSM(
+	txf clienttx.Factory,
+	name string,
+	txBuilder client.TxBuilder,
+	overwriteSig bool,
+	txConfig client.TxConfig,
+) error {
+	hsmCfg, err := GetPKCS11Config()
+	if err != nil {
+		return err
+	}
+
+	address, pubKey, err := GetHSMAddress(hsmCfg, name)
+	if err != nil {
+		return err
+	}
+
+	signerData := authsigning.SignerData{
+		ChainID:       txf.ChainID(),
+		AccountNumber: txf.AccountNumber(),
+		Sequence:      txf.Sequence(),
+		PubKey:        pubKey,
+		Address:       sdk.AccAddress(address).String(),
+	}
+
+	signMode := txf.SignMode()
+
+	// For SIGN_MODE_DIRECT, calling SetSignatures calls setSignerInfos on
+	// TxBuilder under the hood, and SignerInfos is needed to generate the
+	// sign bytes. This is the reason for setting SetSignatures here, with a
+	// nil signature.
+	//
+	// Note: this line is not needed for SIGN_MODE_LEGACY_AMINO, but putting it
+	// also doesn't affect its generated sign bytes, so for code's simplicity
+	// sake, we put it here.
+	sigData := signing.SingleSignatureData{
+		SignMode:  signMode,
+		Signature: nil,
+	}
+	sig := signing.SignatureV2{
+		PubKey:   pubKey,
+		Data:     &sigData,
+		Sequence: txf.Sequence(),
+	}
+
+	var prevSignatures []signing.SignatureV2
+	if !overwriteSig {
+		prevSignatures, err = txBuilder.GetTx().GetSignaturesV2()
+		if err != nil {
+			return err
+		}
+	}
+	// Overwrite or append signer infos.
+	var sigs []signing.SignatureV2
+	if overwriteSig {
+		sigs = []signing.SignatureV2{sig}
+	} else {
+		sigs = append(prevSignatures, sig) //nolint:gocritic
+	}
+	if err := txBuilder.SetSignatures(sigs...); err != nil {
+		return err
+	}
+
+	// Generate the bytes to be signed.
+	bytesToSign, err := txConfig.SignModeHandler().GetSignBytes(signMode, signerData, txBuilder.GetTx())
```

Comment:
It might be a good idea to extract the common part between HSMsign and SIGN into a separate function , and consider a unit test for that

Reply:
Yes, that would be better but unfortunately the Sign function is defined in the sdk and I can't really modify that. Let me know if that is what you meant.